airbyte-integrations/connectors/source-smartsheets/source_smartsheets/source.py

```diff
@@ -0,0 +1,153 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import json
+from datetime import datetime
+from typing import Dict, Generator
+
+import smartsheet
+from airbyte_protocol import (
+    AirbyteCatalog,
+    AirbyteConnectionStatus,
+    AirbyteMessage,
+    AirbyteRecordMessage,
+    AirbyteStream,
+    ConfiguredAirbyteCatalog,
+    Status,
+    Type,
+)
+from base_python import AirbyteLogger, Source
+from smartsheet.smartsheet import Smartsheet
+
+
+# helpers
+def get_prop(col_type: str) -> Dict[str, any]:
+    props = {
+        "TEXT_NUMBER": {"type": "string"},
+        "DATE": {"type": "string", "format": "date"},
+        "DATETIME": {"type": "string", "format": "date-time"},
+    }
+    if col_type in props.keys():
+        return props[col_type]
+    else:  # assume string
+        return props["TEXT_NUMBER"]
+
+
+def get_json_schema(sheet: Smartsheet) -> Dict:
+    column_info = {i["title"]: get_prop(i["type"]) for i in sheet["columns"]}
+    json_schema = {
+        "$schema": "http://json-schema.org/draft-07/schema#",
+        "type": "object",
+        "properties": column_info,
+    }
+    return json_schema
+
+
+# main class definition
+class SourceSmartsheets(Source):
+    def check(self, logger: AirbyteLogger, config: json) -> AirbyteConnectionStatus:
+        try:
+            access_token = config["access_token"]
+            spreadsheet_id = config["spreadsheet_id"] + 1
+
+            smartsheet_client = smartsheet.Smartsheet(access_token)
+            smartsheet_client.errors_as_exceptions(True)
+            smartsheet_client.Sheets.get_sheet(spreadsheet_id)
+
+            return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+        except Exception as e:
+            if isinstance(e, smartsheet.exceptions.ApiError):
+                err = e.error.result
+                code = 404 if err.code == 1006 else err.code
+                reason = f"{err.name}: {code} - {err.message} | Check your spreadsheet ID."
+            else:
+                reason = str(e)
+            logger.error(reason)
```

Comment:
feels like we'd benefit from running `discover` inside this method too to find out if the sheet name is acceptable or not. Ideally by the time we run `discover` there are no "wonky" sheet names or formats that would cause the sync or discovery to fail.

Reply:
can you show me what you mean?

Reply:
the `check` method is usually called before doing any operations with the connector (setting it up as a source in the UI or  before running a sync for example). The idea is to get a sense of whether there are any issues that will prevent interacting with the connector before we run it e.g: invalid credentials, data is structured in a way that a sync can't succeed, etc..

In this case, looking at the discover method [here](https://github.com/airbytehq/airbyte/pull/2880/files#diff-327b64706f0b0b5ce39d8d3f24b12532d82e71b4793c123e824c5e221bc9e0d5R106) I'm seeing that there is a possibility of the sync failing if the sheet is named incorrectly. So why don't we perform that check also in the `check` method so we fail early instead of failing inside the `discover` call, which unlike `check` isn't used to suss out whether there will be any issues ahead? The simplest way to do this would be to call `self.discover` inside the `try` block in the `check` method.

Reply:
ok so that exception message wasn't based on an error I encountered, and in retrospect it doesn't really make sense as a suggestion. the `check` method has always caught invalid configurations so it's a mistake that I left that exception message in there. Do you have any suggestions for what could be helpful as a message in the situation of valid configs and failed stream creation? Also I can still do what you suggested, if still necessary - lmk

Reply:
Ok so I think given this I would suggest: 

1. removing the nested `try..except` block which recommends changing the name to have no spaces or _
2. No changes needed to `check` 


and with that I think we'll be ready to merge!

Reply:
✅